packages/preconfirmation-p2p/crates/service/src/lib.rs

```diff
@@ -0,0 +1,739 @@
+//! High-level async facade over the preconfirmation P2P networking layer.
+//!
+//! This crate owns the `preconfirmation-net` driver and exposes a small, ergonomic
+//! API for sending `NetworkCommand`s and receiving `NetworkEvent`s. Events are fanned out
+//! internally so multiple consumers can subscribe concurrently: pull from `next_event()`
+//! (default subscription), hand a stream to `run_with_handler`, use the blocking helpers,
+//! or call `subscribe()` to obtain additional receivers.
+
+use anyhow::Result;
+use futures::future::poll_fn;
+use std::sync::atomic::{AtomicU64, Ordering};
+use tokio::{
+    sync::{broadcast, mpsc, oneshot},
+    task::JoinHandle,
+};
+
+use libp2p::PeerId;
+pub use preconfirmation_net::{
+    LookaheadResolver, NetworkCommand, NetworkConfig, NetworkDriver, NetworkEvent, PreconfStorage,
+    event::{NetworkError, NetworkErrorKind},
+};
+use preconfirmation_types::{
+    GetCommitmentsByNumberResponse, GetRawTxListResponse, RawTxListGossip, SignedCommitment,
+    validate_raw_txlist_gossip, verify_signed_commitment,
+};
+
+static NEXT_REQUEST_ID: AtomicU64 = AtomicU64::new(1);
+
+/// Application-facing callbacks for P2P events.
+///
+/// Implementors can plug business logic into these methods, allowing the
+/// `P2pService` to hide libp2p details and dispatch high-level events.
+pub trait P2pHandler: Send + Sync + 'static {
+    /// Called when a signed commitment gossip message is received.
+    fn on_signed_commitment(&self, _from: PeerId, _msg: SignedCommitment) {}
+    /// Called when a raw transaction list gossip message is received.
+    fn on_raw_txlist(&self, _from: PeerId, _msg: RawTxListGossip) {}
+    /// Called when a response to a commitment request is received.
+    fn on_commitments_response(
+        &self,
+        _from: PeerId,
+        _msg: GetCommitmentsByNumberResponse,
+        _request_id: Option<u64>,
+    ) {
+    }
+    /// Called when a response to a raw transaction list request is received.
+    fn on_raw_txlist_response(
+        &self,
+        _from: PeerId,
+        _msg: GetRawTxListResponse,
+        _request_id: Option<u64>,
+    ) {
+    }
```

Comment:
And one more point to discuss is, right now the taiko-geth / alethia node doesn't know the txlist (hash) of each block, shall we save that in the node database? or we keep everything in p2p lib's storage. Seems both have some pros and cons.

Reply:
kk let's discuss that later with partners first.